KAFKA-14412: Generalise over RocksDB WriteBatch

nicktelford opened the pull request:
The type hierarchy of RocksDB's `WriteBatch` looks like this:

```
        +---------------------+
        | WriteBatchInterface |
        +---------------------+
                   ^
                   |
        +---------------------+
        |  AbstractWriteBatch |
        +---------------------+
                   ^
                   |
        +----------+----------+
        |                     |
 +------------+    +---------------------+
 | WriteBatch |    | WriteBatchWithIndex |
 +------------+    +---------------------+
```

By switching our `BatchWritingStore` methods from `WriteBatch` to `WriteBatchInterface`, we enable the use of `WriteBatchWithIndex` as well.

nicktelford committed:
KAFKA-14412: Generalise over RocksDB WriteBatch

The type hierarchy of RocksDB's `WriteBatch` looks like this:

```
        +---------------------+
        | WriteBatchInterface |
        +---------------------+
                   ^
                   |
        +---------------------+
        |  AbstractWriteBatch |
        +---------------------+
                   ^
                   |
        +----------+----------+
        |                     |
 +------------+    +---------------------+
 | WriteBatch |    | WriteBatchWithIndex |
 +------------+    +---------------------+
```

By switching our `BatchWritingStore` methods from `WriteBatch` to
`WriteBatchInterface`, we enable the use of `WriteBatchWithIndex` as
well.

nicktelford commented:
@mjsax @ableegoldman @lucasbru @wcarlson5 @bbejeck @vvcephei @guozhangwang

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -572,13 +574,19 @@ public synchronized void flush() {
+        } else if (batch instanceof WriteBatchWithIndex) {
+            db.write(wOptions, (WriteBatchWithIndex) batch);
+        } else {
+            throw new ProcessorStateException("Unknown type of batch " + batch.getClass().getCanonicalName());
```
nit: log an error before we throw the exception

Also, imo this should just be an IllegalStateException instead of a ProcessorStateException

nicktelford replied:
Good point! Done.

nicktelford committed:
Improve error reporting for unknown batch type

We should be using an `IllegalStateException`, and we should log a
message informing the user that this is a bug.

This branch should be unreachable, as both of the possible
implementations of `WriteBatchInterface` are matched in the previous two
branches.

ableegoldman reviewed (approved):
LGTM -- bonus points for the good-looking class hierarchy diagram in the PR description 😄

ableegoldman commented:
Strangely one of the builds failed to compile due to some issue in the upgrade test/smoke client -- seems unrelated but we can't merge with a broken build, so let's retry and hope it goes away 🤞

nicktelford commented:
There was a merge conflict with trunk that has now been resolved. @ableegoldman once the build passes, can we merge this?

ableegoldman commented:
Two of the builds failed again...one of them looks like a failing Connect test bailed out with a sys exit, not sure what happened with the other one, but let's retry it...

nicktelford commented:
I couldn't figure out how to restart the build, but I've rebased against trunk in the hope that some of these failing tests have been addressed upstream.

I'm confident they aren't introduced by this PR.

ableegoldman commented:
You need to log in to restart the build, so it might be that only committers can trigger a retry...which is annoying (though I guess I kind of understand, don't want some random person to waste money & compute by spamming retries. We've had weirdos spam kafka before so it's not unthinkable 🤷‍♀️ )

Anyways I replayed it but looks like now there's some issue with Jenkins and/or git itself, so I replayed it again just now. FWIW I really don't think this PR is breaking anything, but I just can't merge with a completely broken build. Hopefully this time it goes through 🤞 
![Screenshot 2023-12-14 at 12 41 17 PM](https://github.com/apache/kafka/assets/4843099/39f09f1e-2d3b-446b-8a40-782d7474d154)

nicktelford commented:
The build still appears to be failing, this time with an error from git about updating the HEAD... It doesnt look like other PRs are affected by this, so I'm guessing the build is cached per-PR and something has borked the cached build for this PR.

Shall I just close this PR and open a new one from the same branch? Might give it a fresh build that can then succeed.

ableegoldman commented:
I honestly don't know...maybe try doing a rebase from trunk and force pushing? That can often clear up this sort of thing. Ideally we don't have to abandon this PR for a fresh one (just to avoid losing history/review/comments), but that can be a last resort I suppose
